File: tools/pq-ui/pq1/motion.py
"""PQ1 motion vocabulary — easing, timing and input-gesture constants, follower chain, chevron hint.

Screen-to-screen choreography lives in pq1.flow.Sim (and pq1.driver for
interactive playback); this module holds the shared
building blocks so every screen moves the same way.
"""
import math


# ---------------------------------------------------------------- easing ----
def clamp01(t):
    return max(0.0, min(1.0, t))


def ease(t):
    """cubic in-out"""
    return 4 * t * t * t if t < 0.5 else 1 - (-2 * t + 2) ** 3 / 2


def ease_out(t):
    """cubic ease-out — press feedback, hold snap-back, scripted fades"""
    return 1 - (1 - t) ** 3


def back_out(t):
    c1 = 1.70158
    c3 = c1 + 1
    return 1 + c3 * (t - 1) ** 3 + c1 * (t - 1) ** 2


def lerp(a, b, t):
    return a + (b - a) * t


# ----------------------------------------------------------------- spring ---
class Spring:
    """Damped spring in Apple's designer parameterization.

    response — seconds; how quickly the value approaches the target (not a
    duration: the settle time emerges from the physics). damping — damping
    ratio; 1.0 = critically damped, no overshoot, the default for every UI
    value here (two buttons carry no momentum, so nothing should bounce).

    step() advances with the closed-form solution of the damped linear ODE
    over the frame's dt, so it is exact at any frame rate — the same
    dt-stepping idiom as FollowerChain.step. retarget() only moves the
    goalpost: position and velocity always carry over, which is what makes
    spring motion interruptible mid-flight.

    A landed spring (within SNAP_EPS of its target, velocity under SNAP_VEL)
    snaps exactly onto the target and sleeps: step() early-returns until
    retarget() moves the goal. Code that pokes .value/.target directly must
    leave them consistent (at target, zero velocity) or clear .asleep --
    retarget() handles this for you.
    """

    SNAP_EPS = 1e-3   # position snap tolerance (px or alpha units)
    SNAP_VEL = 1e-2   # velocity snap tolerance (units per second)

    def __init__(self, value=0.0, response=0.4, damping=1.0):
        self.value = float(value)
        self.velocity = 0.0  # units per second
        self.target = float(value)
        self.response = response
        self.damping = min(1.0, max(0.05, damping))  # overdamped never wanted in UI
        self.asleep = False

    def retarget(self, target):
        """move the goalpost; the live pose and velocity are kept"""
        t = float(target)
        if t != self.target:
            self.asleep = False  # a real goal change always wakes the spring
        self.target = t

    def settled(self, tol=0.5):
        return abs(self.value - self.target) < tol and abs(self.velocity) < tol * 10

    def step(self, dt_ms):
        if self.asleep:
            return self.value
        dt = dt_ms / 1000.0
        if dt <= 0:
            return self.value
        w = 2 * math.pi / self.response
        z = self.damping
        x = self.value - self.target
        v = self.velocity
        if z >= 1.0:  # critically damped: x(t) = (x0 + B t) e^{-wt}
            B = v + w * x
            e = math.exp(-w * dt)
            self.value = self.target + (x + B * dt) * e
            self.velocity = (B - w * (x + B * dt)) * e
        else:  # underdamped: decaying oscillation at wd
            wd = w * math.sqrt(1 - z * z)
            B = (v + z * w * x) / wd
            e = math.exp(-z * w * dt)
            c, s = math.cos(wd * dt), math.sin(wd * dt)
            self.value = self.target + e * (x * c + B * s)
            self.velocity = e * ((B * wd - z * w * x) * c - (x * wd + z * w * B) * s)
        if (abs(self.value - self.target) < self.SNAP_EPS
                and abs(self.velocity) < self.SNAP_VEL):
            self.value = self.target  # settle-snap: land exactly, then sleep
            self.velocity = 0.0
            self.asleep = True
        return self.value


# spring profiles — one pace per context for every transition value: circle
# travel/radius, glyph+chevron morph AND the per-screen text alpha crossfade
# (text arrives with the circle rather than popping in ahead of it)
NAV = dict(response=0.40, damping=1.0)    # hardware button pace (Apple "move")
KIOSK = dict(response=0.55, damping=1.0)  # demo-loop pace; the Sim default


def spring_travel(t_ms, profile=KIOSK):
    """progress 0 -> 1 of a critically damped spring released from rest at
    t = 0 — Spring.step's closed form, 1 - (1 + wt) e^-wt, as a pure function
    of t: a film's circle travels exactly like a flow transition's"""
    if t_ms <= 0:
        return 0.0
    wt = 2 * math.pi / profile["response"] * t_ms / 1000.0
    return 1.0 - (1.0 + wt) * math.exp(-wt)


ENTER_MS = 800  # a film's side entrance (off-panel -> centre) at KIOSK pace:
                # the ~274 px trip has settled to 0.3 px by here, then snaps


# ------------------------------------------------------------ timing (ms) ---
TEXT_IN_DELAY_MS = 150  # incoming text starts its alpha spring this long after
                        # the leg begins: the circle leads, the text lands just
                        # after it (outgoing text starts fading immediately)
FADE_MS = 180           # pre-spring serial fade; survives as a span estimate
MOVE_MS = 900           # pre-spring travel time; MOVE_MS + 2*FADE_MS still
                        # bounds a transition (springs settle well inside it)
HERO_DWELL = 5000       # hero screens idle time (one full sweep)
DETAIL_DWELL = 4100     # detail screens idle time
STATUS_DWELL = 8600     # the qubit status animation's duration (loading +
                        # resolve + hold); status screens now dwell for their
                        # own animation's duration — status.py asserts the
                        # film ("qubit", the done-ending default) still equals
                        # this documented value (the film-less cancel
                        # "resolve" dwells 2850 ms)

# -------------------------------------------------------------- input (ms) --
# Two-button gesture timing — spec'd in DESIGN.md § Input; consumed by the
# reference driver (pq1/driver.py, the bench player's grammar). Device
# firmware re-implements the grammar natively on-device from the same spec.
PRESS_FEEDBACK_MS = 120  # press-down acknowledgment nudge on the pressed-side chevron
TAP_MAX_MS = 250         # released under this = tap; held past it = hold begins
DOUBLE_TAP_MS = 250      # second press within this converts a tap (entry contexts only)
CHORD_MS = 150           # both buttons: a press on one side within this of the other
                         # side's press is the chord (ENTER on an entry)
HOLD_COMMIT_MS = 2000    # linear progress fill; the hold action fires ONLY at
                         # completion — releasing a moment earlier does nothing
HOLD_SNAPBACK_MS = 200   # ease-out snap-back when a hold is released early

LEFT, RIGHT = "left", "right"   # the two buttons (pq1.driver re-exports them)


def hold_fill(t_since_press, t_release=None):
    """Hold progress 0..1 — DESIGN.md § Input made code.

    Flat 0 until TAP_MAX_MS (a tap never flashes a partial arc — the fill's
    appearance IS the "this became a hold" signal), then linear to 1 at
    HOLD_COMMIT_MS from press-down, exactly when the hold action fires.
    t_release (ms since press at which the button came up, before commit)
    freezes the fill at its release value and snaps it back with ease_out
    over HOLD_SNAPBACK_MS. A pure function of time, so any frame is seekable.
    The visual is components.hold_flood; pq1.flow.Sim and the library's
    hold_to_confirm screen both draw through it."""
    def k_at(t):
        return clamp01((t - TAP_MAX_MS) / (HOLD_COMMIT_MS - TAP_MAX_MS))
    if t_release is None:
        return k_at(t_since_press)
    k = k_at(min(t_since_press, t_release))
    return k * (1 - ease_out(clamp01((t_since_press - t_release)
                                     / HOLD_SNAPBACK_MS)))


def hold_fill_ms(k):
    """inverse of hold_fill's rising edge: ms since press at which the fill
    reaches k (hold_fill_ms(1.0) == HOLD_COMMIT_MS)"""
    return TAP_MAX_MS + k * (HOLD_COMMIT_MS - TAP_MAX_MS)

# -------------------------------------------------------------- idle sweep --
SWEEP_DELAY_MS = 1000   # centred hold before the sweep starts
SWEEP_PERIOD_MS = 5000  # one full left-right cycle
SWEEP_AMP = 95          # sweep amplitude (px)
OSC_TAU = 180.0         # easing constant chasing the sweep target

# --------------------------------------------------------- follower chain ---
TRAIL_COUNT = 5         # follower circles in the streak chain
MAX_GAP = 30.0          # px cap between consecutive links
CHAIN_TAU = 60.0        # ms easing constant per link
CHAIN_TAU_IDLE = 150.0  # slower chase during the idle sweep -> more separation


def tau_k(dt_ms, tau_ms):
    """per-frame coefficient of the frame-rate-independent tau-chase"""
    return 1.0 - math.exp(-dt_ms / tau_ms)


def tau_chase(current, target, dt_ms, tau_ms):
    """one tau-chase step: move current toward target by 1 - exp(-dt/tau) --
    THE ambient-motion idiom (idle sweep, follower chain); DESIGN.md S Motion"""
    return current + (target - current) * tau_k(dt_ms, tau_ms)


class FollowerChain:
    """Eased chain of follower points trailing a moving head.

    Call step() once per frame with the head position; read .points
    (nearest-first) to draw the streak.
    """

    def __init__(self, x, y, count=TRAIL_COUNT, max_gap=MAX_GAP):
        self.points = [dict(x=x, y=y) for _ in range(count)]
        self.max_gap = max_gap

    def step(self, hx, hy, dt, tau=CHAIN_TAU):
        for p in self.points:  # at rest (head + links coincident)? skip it all
            if abs(p["x"] - hx) > 1e-3 or abs(p["y"] - hy) > 1e-3:
                break
        else:
            return
        k = tau_k(dt, tau)  # hoisted: one exp per call, not one per link
        gap = self.max_gap
        px, py = hx, hy
        for p in self.points:
            x = p["x"] + (px - p["x"]) * k
            y = p["y"] + (py - p["y"]) * k
            dx, dy = x - px, y - py
            if dx * dx + dy * dy > gap * gap:  # sqrt only when clamping
                d = math.hypot(dx, dy)
                x = px + dx / d * gap
                y = py + dy / d * gap
            p["x"], p["y"] = x, y
            px, py = x, y


# ---------------------------------------------------------- view-more hint --
BAND_SWAP_MS = 5000   # confirm band: each message holds this long
BAND_FADE_MS = 300    # ... and fades away / back in over this
CONFIRM_DWELL = 2 * BAND_SWAP_MS   # demo dwell: both band messages play


def confirm_band(idle_ms):
    """Confirm-screen band alternation -> (a_more, a_back): the "OR VIEW
    MORE" unit holds for BAND_SWAP_MS, fades away over BAND_FADE_MS, and
    "TO GO BACK" fades in for the next slot — a sequential swap every 5 s
    while the screen idles. Alphas multiply the screen's own alpha."""
    p = idle_ms % (2 * BAND_SWAP_MS)
    t = p % BAND_SWAP_MS
    a = (ease_out(clamp01(t / BAND_FADE_MS))
         * (1 - ease_out(clamp01((t - (BAND_SWAP_MS - BAND_FADE_MS))
                                 / BAND_FADE_MS))))
    return (a, 0.0) if p < BAND_SWAP_MS else (0.0, a)


# ------------------------------------------------------------- page flip --
PAGE_SWAP_MS = DETAIL_DWELL   # a paged detail: each page holds one detail dwell
PAGE_FADE_MS = BAND_FADE_MS   # ... the showing page fades away / the next in over this


def page_flip(t_ms):
    """Paged-detail swap envelope -> (a_out, a_in) for a flip that began
    t_ms ago (DESIGN.md § Motion, Page flip): the outgoing page fades away
    over PAGE_FADE_MS on the ease-out curve, THEN the incoming page fades
    in over the next PAGE_FADE_MS on the ease curve — a sequential swap on
    the confirm band's timing; the pager number switches between the two
    phases. Alphas multiply the screen's own alpha."""
    if t_ms < PAGE_FADE_MS:
        return 1.0 - ease_out(clamp01(t_ms / PAGE_FADE_MS)), 0.0
    return 0.0, ease(clamp01((t_ms - PAGE_FADE_MS) / PAGE_FADE_MS))


BUSY_PULSE_MS = 2000   # a loading caption breathes: one fade-in + fade-out a cycle


def busy_pulse(u):
    """the loading caption's breathing alpha over one cycle's unit progress
    u in [0, 1]: a raised cosine — dark at 0, full at 1/2, dark again at 1 —
    so the text fades in and out, slowly, for as long as the loading runs
    (status.StatusAnim.draw_busy fits whole cycles into the busy window;
    user rule, Sep 2026: text during a loading animation pulsates, and only
    once the circles are already going round)"""
    return 0.5 - 0.5 * math.cos(2 * math.pi * clamp01(u))


CHEV_HINT_PERIOD_MS = 3600   # hero hint cycle; confirm screens run the
                             # same envelope on the band beat (BAND_SWAP_MS)


def chevron_hint(idle_ms, period_ms=CHEV_HINT_PERIOD_MS):
    """Chevron hint envelope: (hint_up, y_offset).

    hint_up 0..1 rotates the corner chevrons toward "up"; y_offset bobs them
    while fully up. Starts 1.4 s into the idle and repeats every period_ms —
    3.6 s on hero screens; confirm screens pass BAND_SWAP_MS so the pointing
    gesture lands on the band's 5 s beat (their chevrons already rest up,
    so only the bob shows).
    """
    hint_up = 0.0
    y_off = 0.0
    if idle_ms > 1400:
        p = (idle_ms - 1400) % period_ms
        if p < 350:
            hint_up = ease(p / 350)
        elif p < 1550:
            hint_up = 1.0
            y_off = -4 * math.sin(math.pi * (p - 350) / 1200)
        elif p < 1900:
            hint_up = 1 - ease((p - 1550) / 350)
    return hint_up, y_off


# ------------------------------------------------------------ accent curves --
# One-shot expressive curves for verdict/notice screens (screens/ library).
# All are pure functions of normalized progress t in [0, 1] (clamp before
# calling). One overshoot is sanctioned in PQ1 — back_out (the status
# flash pop), a celebration, never navigation. The verdict icon ENTRANCE
# never overshoots: it fades in and rises 0.97 -> 1 (arrive), both
# ease_out, within ARRIVE_MS (user rule, Sep 2026).

ARRIVE_MS = 300      # the verdict sign's entrance (fade + arrive) — never longer
ARRIVE_FROM = 0.97   # the entrance scale starts here and settles on 1.0


def arrive(t):
    """verdict icon entrance scale — a near-rest rise ARRIVE_FROM -> 1 on
    ease_out; pair it with an ease_out fade (VerdictAnim.entrance)"""
    return lerp(ARRIVE_FROM, 1.0, ease_out(t))


def pop(t):
    """springy overshoot with one designed wobble — an opt-in celebration
    accent, no longer any entrance (the verdict sign arrives, above)"""
    return 1 - math.exp(-5 * t) * math.cos(2.2 * math.pi * t)


def attention_pulse(t):
    """decaying scale pulse (1.0-centred) — draws the eye to a settled icon"""
    return 1 + 0.09 * math.exp(-3 * t) * abs(math.sin(2.5 * math.pi * t))


def heartbeat(t, pumps=3.0, amp=0.22, decay=4.0):
    """decaying pump train (1.0-centred) — last-attempt heart"""
    return 1 + amp * math.exp(-decay * t) * abs(math.sin(pumps * math.pi * t))


def shake(t, cycles=2.5):
    """decaying error shake, unit amplitude — multiply by the px excursion"""
    return math.sin(2 * math.pi * cycles * t) * (1 - t)


def wobble(t, decay=3.0):
    """decaying settle oscillation, unit amplitude"""
    return math.exp(-decay * t) * math.sin(2 * math.pi * t)


def recoil(t):
    """mechanical click recoil, unit amplitude — rises and dies within t"""
    return math.sin(math.pi * t) * (1 - t)


def freewheel(t, k=4.2):
    """coasting spin-down: fast start, exponential decay to a stop at t=1"""
    return (1 - math.exp(-k * t)) / (1 - math.exp(-k))


def decel(t, p=2.6):
    """simple power deceleration 1-(1-t)^p — tumbles, reels, ring growth"""
    return 1 - (1 - t) ** p


# every one-shot accent phase (ring stagger, click, bounce) must span at
# least this long — two panel frames at the NV3007's 14 fps
VERDICT_ACCENT_MIN_MS = 145

File: pqsigner-ui-px/src/motion.rs
//! Motion vocabulary — a fixed-point port of `tools/pq-ui/pq1/motion.py`.
//!
//! Springs drive every transition (Apple's `response` / `damping`
//! parameterisation, damping 1.0 everywhere: two buttons carry no momentum,
//! so navigation never overshoots). A landed spring snaps onto its target
//! and sleeps — zero cost per frame until the next retarget.
//!
//! Timing constants are the design's (§ Input, § Motion); the accent curves
//! are the named one-shots verdict screens use.

use crate::fixed::{clamp01, exp_neg_q16, lerp, mul_q16, Q16, ONE_Q16};

// ---- gesture timing (DESIGN.md § Input) ----------------------------------
/// Press-down acknowledgment: the pressed-side chevron nudges.
pub const PRESS_FEEDBACK_MS: u32 = 120;
/// Released under this = a tap; held past it = a hold begins.
/// A release up to this long after the press is a tap; longer is an aborted
/// hold. The PQ-UI reference says 250 ms; on the pq1's physical switches a
/// deliberate press runs 250–400 ms (EVT #1, 2026-09-22: single "clicks"
/// nudged the hero and did nothing, only a fast double-click advanced), so
/// the device uses the 500 ms the legacy button driver already validated.
pub const TAP_MAX_MS: u32 = 500;
/// A second press within this of a tap converts it (entry contexts only).
pub const DOUBLE_TAP_MS: u32 = 250;
/// A press on the other side within this is the chord.
pub const CHORD_MS: u32 = 150;
/// The hold fires exactly here; earlier release does nothing.
pub const HOLD_COMMIT_MS: u32 = 2000;
/// Early-release snap-back of the fill.
pub const HOLD_SNAPBACK_MS: u32 = 200;

// ---- choreography ---------------------------------------------------------
/// Incoming text is released this long after the circle starts moving.
pub const TEXT_IN_DELAY_MS: u32 = 150;
/// Idle sweep: centred hold, then a 5 s left-right cycle of ±95 px.
pub const SWEEP_HOLD_MS: u32 = 1000;
pub const SWEEP_PERIOD_MS: u32 = 5000;
pub const SWEEP_AMPLITUDE_PX: i32 = 95;
/// τ-chase time constants.
pub const OSC_TAU_MS: u32 = 180;
pub const CHAIN_TAU_MS: u32 = 60;
pub const CHAIN_TAU_IDLE_MS: u32 = 150;
pub const CHAIN_GAP_CAP_PX: i32 = 30;
/// Chevron hint: starts 1.4 s into the idle, repeats every 3.6 s.
pub const HINT_START_MS: u32 = 1400;
pub const HINT_PERIOD_MS: u32 = 3600;
pub const HINT_ROTATE_MS: u32 = 350;
pub const HINT_BOB_MS: u32 = 1200;
pub const HINT_BOB_PX: i32 = 4;
/// Confirm band: OR VIEW MORE ▸ / ◂ TO GO BACK swap every 5 s over 300 ms.
pub const BAND_SWAP_MS: u32 = 5000;
pub const BAND_FADE_MS: u32 = 300;
/// Page flip: sequential fade out then in, 300 ms each.
pub const PAGE_FADE_MS: u32 = 300;
/// Verdict entrance.
pub const ARRIVE_MS: u32 = 300;
/// Every one-shot accent phase spans at least two panel frames at 14 fps.
pub const VERDICT_ACCENT_MIN_MS: u32 = 145;
/// Hold overlay film alpha (30 %).
pub const HOLD_OVERLAY_A8: u8 = 77;

/// Spring profile: `w = 2π / response` in Q16 per millisecond.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Profile {
    pub w_q16_per_ms: Q16,
}

impl Profile {
    /// Hardware button pace (`response` 0.40 s).
    pub const NAV: Self = Self {
        // 2π / 400 ms = 0.015708 → Q16 = 1029.4
        w_q16_per_ms: 1029,
    };
    /// Demo loops (`response` 0.55 s); not used on the device.
    pub const KIOSK: Self = Self {
        // 2π / 550 ms = 0.011424 → Q16 = 748.7
        w_q16_per_ms: 749,
    };
}

/// A critically damped spring on a Q16 value.
#[derive(Clone, Copy, Debug)]
pub struct Spring {
    pub value: Q16,
    /// Q16 per millisecond.
    pub velocity: Q16,
    pub target: Q16,
    pub profile: Profile,
    asleep: bool,
}

/// Snap tolerance: 1/64 of a unit (a sixty-fourth of a pixel, 4/255 of an
/// alpha) — integer truncation stalls the closed form a few Q16 units short
/// of the target, well inside this.
const SNAP_EPS: Q16 = 1 << 10;
/// Velocity sleep tolerance: ~1e-4 per ms (8 Q16/ms) — well under a pixel per second.
const SNAP_VEL: Q16 = 64;

impl Spring {
    #[must_use]
    pub const fn at(value: Q16, profile: Profile) -> Self {
        Self {
            value,
            velocity: 0,
            target: value,
            profile,
            asleep: true,
        }
    }

    /// Retarget from the live pose (velocity carries over).
    pub fn retarget(&mut self, target: Q16) {
        self.target = target;
        self.asleep = false;
    }

    /// Snap onto a value and sleep.
    pub fn snap(&mut self, value: Q16) {
        self.value = value;
        self.target = value;
        self.velocity = 0;
        self.asleep = true;
    }

    #[must_use]
    pub fn settled(&self) -> bool {
        self.asleep
    }

    /// Advance by `dt` ms (capped at 100 by the caller).
    pub fn step(&mut self, dt_ms: u32) {
        if self.asleep || dt_ms == 0 {
            return;
        }
        let w = i64::from(self.profile.w_q16_per_ms);
        let dt = i64::from(dt_ms.min(100));
        let x = i64::from(self.value) - i64::from(self.target);
        let v = i64::from(self.velocity);
        // Critically damped: d(t) = (A + B t) e^{-w t}, A = x, B = v + w x.
        let b = v + ((w * x) >> 16);
        let e = i64::from(exp_neg_q16(((w * dt) as i32).min(8 << 16)));
        let disp = x + b * dt;
        let nx = (disp * e) >> 16;
        let nv = ((b - ((w * disp) >> 16)) * e) >> 16;
        self.value = (i64::from(self.target) + nx) as Q16;
        self.velocity = nv as Q16;
        if (self.value - self.target).abs() <= SNAP_EPS && self.velocity.abs() <= SNAP_VEL {
            self.snap(self.target);
        }
    }
}

// ---- named curves ---------------------------------------------------------

/// Cubic ease in-out, `u` in Q16.
#[must_use]
pub fn ease(u: Q16) -> Q16 {
    let u = clamp01(u);
    if u < ONE_Q16 / 2 {
        4 * mul_q16(mul_q16(u, u), u)
    } else {
        let f = 2 * u - 2 * ONE_Q16;
        ONE_Q16 + mul_q16(mul_q16(f, f), f) / 2
    }
}

/// Cubic ease out.
#[must_use]
pub fn ease_out(u: Q16) -> Q16 {
    let f = clamp01(u) - ONE_Q16;
    ONE_Q16 + mul_q16(mul_q16(f, f), f)
}

/// The one sanctioned overshoot (status flash pop): back-out with s = 1.70158.
#[must_use]
pub fn back_out(u: Q16) -> Q16 {
    let f = clamp01(u) - ONE_Q16;
    let s: Q16 = 111_529; // 1.70158
    // 1 + f² ((s+1) f + s)
    ONE_Q16 + mul_q16(mul_q16(f, f), mul_q16(s + ONE_Q16, f) + s)
}

/// Verdict entrance: (alpha, scale) both ease-out, scale 0.97 → 1.
#[must_use]
pub fn arrive(u: Q16) -> (Q16, Q16) {
    let e = ease_out(u);
    (e, lerp(63_570, ONE_Q16, e))
}

/// Progress `[0, 1]` of `t` within `[start, start + dur]`.
#[must_use]
pub fn phase(t_ms: u32, start_ms: u32, dur_ms: u32) -> Q16 {
    if dur_ms == 0 || t_ms <= start_ms {
        return 0;
    }
    let el = t_ms - start_ms;
    if el >= dur_ms {
        return ONE_Q16;
    }
    ((u64::from(el) << 16) / u64::from(dur_ms)) as Q16
}

/// Hold fill level for a press that started at t 0: flat until
/// `TAP_MAX_MS`, then linear to 1 at `HOLD_COMMIT_MS`.
#[must_use]
pub fn hold_fill(held_ms: u32) -> Q16 {
    if held_ms <= TAP_MAX_MS {
        return 0;
    }
    phase(held_ms, TAP_MAX_MS, HOLD_COMMIT_MS - TAP_MAX_MS)
}

/// Snap-back from `level` starting at release: ease-out drain over 200 ms.
#[must_use]
pub fn hold_snapback(level: Q16, since_release_ms: u32) -> Q16 {
    let u = phase(since_release_ms, 0, HOLD_SNAPBACK_MS);
    mul_q16(level, ONE_Q16 - ease_out(u))
}

/// Frame-rate-independent τ-chase factor `1 − e^{−dt/τ}` in Q16.
#[must_use]
pub fn tau_k(dt_ms: u32, tau_ms: u32) -> Q16 {
    if tau_ms == 0 {
        return ONE_Q16;
    }
    let t = ((u64::from(dt_ms.min(1000)) << 16) / u64::from(tau_ms)) as Q16;
    ONE_Q16 - exp_neg_q16(t.min(8 << 16))
}

/// Chevron hint pose at idle time `t`: (rotate-up amount Q16, bob px).
/// Starts at 1.4 s, repeats every 3.6 s: 350 ms rotate up, 1200 ms bob
/// (−4 px sine), 350 ms return.
#[must_use]
pub fn chevron_hint(idle_ms: u32) -> (Q16, i32) {
    if idle_ms < HINT_START_MS {
        return (0, 0);
    }
    let t = (idle_ms - HINT_START_MS) % HINT_PERIOD_MS;
    if t < HINT_ROTATE_MS {
        (ease(phase(t, 0, HINT_ROTATE_MS)), 0)
    } else if t < HINT_ROTATE_MS + HINT_BOB_MS {
        let u = phase(t, HINT_ROTATE_MS, HINT_BOB_MS);
        // −4 px · sin(π u): half a turn.
        let s = crate::fixed::sin_turns_q16(u / 2);
        (ONE_Q16, -(((i64::from(s) * i64::from(HINT_BOB_PX)) + (1 << 15)) >> 16) as i32)
    } else if t < 2 * HINT_ROTATE_MS + HINT_BOB_MS {
        (ONE_Q16 - ease(phase(t, HINT_ROTATE_MS + HINT_BOB_MS, HINT_ROTATE_MS)), 0)
    } else {
        (0, 0)
    }
}

/// Confirm band state at settled time `t`: (which message: 0 = OR VIEW MORE,
/// 1 = TO GO BACK; its alpha Q16). Sequential 300 ms swap every 5 s.
#[must_use]
pub fn confirm_band(settled_ms: u32) -> (u8, Q16) {
    let cycle = 2 * BAND_SWAP_MS;
    let t = settled_ms % cycle;
    let (which, local) = if t < BAND_SWAP_MS { (0u8, t) } else { (1u8, t - BAND_SWAP_MS) };
    // Fade out over the last 300 ms of the slot, fade in over the first 300.
    let alpha = if local < BAND_FADE_MS {
        ease(phase(local, 0, BAND_FADE_MS))
    } else if local > BAND_SWAP_MS - BAND_FADE_MS {
        ONE_Q16 - ease_out(phase(local, BAND_SWAP_MS - BAND_FADE_MS, BAND_FADE_MS))
    } else {
        ONE_Q16
    };
    (which, alpha)
}

/// Page flip alphas at `t` since the flip started: (outgoing, incoming) —
/// out fades over 300 ms (ease-out), THEN in over 300 ms (ease).
#[must_use]
pub fn page_flip(t_ms: u32) -> (Q16, Q16) {
    if t_ms < PAGE_FADE_MS {
        (ONE_Q16 - ease_out(phase(t_ms, 0, PAGE_FADE_MS)), 0)
    } else {
        (0, ease(phase(t_ms, PAGE_FADE_MS, PAGE_FADE_MS)))
    }
}

/// Idle sweep target x-offset (px) at idle time `t`: 0 for the first second,
/// then a ±95 px sine with a 5 s period.
#[must_use]
pub fn sweep_offset(idle_ms: u32) -> i32 {
    if idle_ms < SWEEP_HOLD_MS {
        return 0;
    }
    let t = (idle_ms - SWEEP_HOLD_MS) % SWEEP_PERIOD_MS;
    let turns = ((u64::from(t) << 16) / u64::from(SWEEP_PERIOD_MS)) as Q16;
    let s = crate::fixed::sin_turns_q16(turns);
    (((i64::from(s) * i64::from(SWEEP_AMPLITUDE_PX)) + (1 << 15)) >> 16) as i32
}

#[cfg(test)]
mod tests {
    use super::*;

    fn f(q: Q16) -> f64 {
        f64::from(q) / 65536.0
    }

    #[test]
    fn spring_converges_without_overshoot_and_sleeps() {
        let mut s = Spring::at(0, Profile::NAV);
        s.retarget(100 << 16);
        let mut last = 0;
        let mut t = 0;
        while !s.settled() && t < 5000 {
            s.step(16);
            t += 16;
            assert!(s.value >= last - 64, "monotonic (±1e-3): {} then {}", last, s.value);
            assert!(s.value <= (100 << 16), "no overshoot");
            last = s.value;
        }
        assert!(s.settled(), "settles");
        assert_eq!(s.value, 100 << 16);
        // NAV settles in roughly half a second (design: ~500 ms per leg).
        assert!((400..=1100).contains(&t), "settled at {t} ms");
    }

    #[test]
    fn spring_closed_form_matches_python_reference() {
        // motion.py: response 0.40 → after 100 ms from x=1, v=0:
        // d = (1 + w·0.1) e^{-w·0.1}, w = 2π/0.4 = 15.708 → (2.5708)(0.2079) = 0.5345
        let mut s = Spring::at(ONE_Q16, Profile::NAV);
        s.retarget(0);
        s.step(100);
        assert!((f(s.value) - 0.5345).abs() < 0.01, "{}", f(s.value));
    }

    #[test]
    fn curves_have_the_right_endpoints() {
        assert_eq!(ease(0), 0);
        assert!((f(ease(ONE_Q16)) - 1.0).abs() < 1e-3);
        assert!((f(ease(ONE_Q16 / 2)) - 0.5).abs() < 1e-3);
        assert_eq!(ease_out(0), 0);
        assert!((f(ease_out(ONE_Q16)) - 1.0).abs() < 1e-3);
        assert!(f(back_out(ONE_Q16 * 7 / 10)) > 1.0, "back-out overshoots");
        assert!((f(back_out(ONE_Q16)) - 1.0).abs() < 1e-3);
        let (a, sc) = arrive(ONE_Q16);
        assert!((f(a) - 1.0).abs() < 1e-3 && (f(sc) - 1.0).abs() < 1e-3);
        assert!((f(arrive(0).1) - 0.97).abs() < 1e-3);
    }

    #[test]
    fn hold_fill_timing() {
        assert_eq!(hold_fill(0), 0);
        assert_eq!(hold_fill(TAP_MAX_MS), 0);
        assert!((f(hold_fill((TAP_MAX_MS + HOLD_COMMIT_MS) / 2)) - 0.5).abs() < 0.01);
        assert_eq!(hold_fill(HOLD_COMMIT_MS), ONE_Q16);
        assert_eq!(hold_snapback(ONE_Q16, 0), ONE_Q16);
        assert_eq!(hold_snapback(ONE_Q16, HOLD_SNAPBACK_MS), 0);
    }

    #[test]
    fn hint_band_flip_and_sweep_cycles() {
        assert_eq!(chevron_hint(0), (0, 0));
        let (r, _) = chevron_hint(HINT_START_MS + HINT_ROTATE_MS + 10);
        assert_eq!(r, ONE_Q16);
        let (_, bob) = chevron_hint(HINT_START_MS + HINT_ROTATE_MS + HINT_BOB_MS / 2);
        assert_eq!(bob, -HINT_BOB_PX);
        assert_eq!(chevron_hint(HINT_START_MS + HINT_PERIOD_MS - 10), (0, 0));

        assert_eq!(confirm_band(2500), (0, ONE_Q16));
        assert_eq!(confirm_band(7500), (1, ONE_Q16));
        assert!(confirm_band(BAND_SWAP_MS - 1).1 < ONE_Q16 / 10);

        assert_eq!(page_flip(0).0, ONE_Q16);
        assert_eq!(page_flip(PAGE_FADE_MS), (0, 0));
        assert_eq!(page_flip(2 * PAGE_FADE_MS).1, ONE_Q16);

        assert_eq!(sweep_offset(500), 0);
        assert_eq!(sweep_offset(SWEEP_HOLD_MS + SWEEP_PERIOD_MS / 4), SWEEP_AMPLITUDE_PX);
        assert_eq!(sweep_offset(SWEEP_HOLD_MS + 3 * SWEEP_PERIOD_MS / 4), -SWEEP_AMPLITUDE_PX);
        assert!((f(tau_k(180, 180)) - 0.632).abs() < 0.01);
    }
}
